src/cmds/mem/infomem.c

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file
+ * @brief
+ *
+ * @author  Evgenij Bogdanov
+ * @date    09.07.2018
+ */
+
+#include <stdio.h>
+
+#include <mem/phymem.h>
+#include <mem/vmem/vmem_alloc.h>
+#include <mem/vmem.h>
+#include <drivers/common/memory.h>
+#include <framework/mod/options.h>
+
+static void print_vmem(struct page_allocator *pgd_allocator,
+		struct page_allocator *pmd_allocator, struct page_allocator *pte_allocator) {
+	int is_on_vmem = vmem_mmu_enabled();
+
+	printf("VIRTUAL MEMORY:\n");
+	printf("vmem is %s\n", is_on_vmem ? "ON" : "OFF");
+
+	if (is_on_vmem) {
+		if (pte_allocator != NULL) {
+			printf("max pages count / free - %llu / %zu\n",
+					(unsigned long long) (pgd_allocator->pages_n * pmd_allocator->pages_n) * pte_allocator->pages_n,
```

Comment:
I am not sure the total count of pages is a multiple of these three values because of the following reasons. 1. These allocators do not allocate the pages, they allocate tables for the pages. That is, each single page allocated by the allocator, represents a table with multiple entries, which then used to get physical address and page's attributes. 2. These allocator do not form a tree: you cannot allocate  pmd_allocator->pages_n for every pgd entry, you can allocate only pmd_allocator->pages_n at all.

I'd propose to print something like this:
```
printf("PGD tables count / free - %llu / %zu\n", pgd_allocator->pages_n,
          pgd_allocator->free / pgd_allocator->page_size);
printf("PMD tables count / free - %llu / %zu\n", pmd_allocator->pages_n,
          pmd_allocator->free / pmd_allocator->page_size);
if (pte_allocator) {
    printf("PTE tables count / free - %llu / %zu\n", pte_allocator->pages_n,
             pte_allocator->free / pte_allocator->page_size)
} else {
    printf("PTE do not available in this configuration\n");
}
```

P.S.
If and when you have time you can investigate virtual memory tables structure by looking at src/mem/vmem and especially at vmem_mapper.c

Reply:
Yes, of course, there is a way to calculate the actual number of the mapped pages, and you can investigate how to do that further.

Reply:
Ok, thank you, I'm not right understand the allocators meaning.
I fixed output.

Also there is a somthing strange with PGD tables: number of free tables is more than number of tables at all. (2048 / 2049 - all / free).

Reply:
Well, I tried infomem on arm/qemu and reproduced the bug. After looking a bit more in depth, gdb shows me that there were happened 2 page_alloc's and 3 page_free on pgd_allocator. The last page_free worked well and incremented allocator->free up to the maximum available space + page_size. Apparently, it's bug in bimask.c and somewhere else (in there who try free non-allocated page). So the output of your command is well.

Thank you, I will add an issue.

Reply:
I created the issue - https://github.com/embox/embox/issues/1185